docs/architecture/adr-018-network-upgrades.md

```diff
@@ -29,40 +30,48 @@ One of the main difficulties of social upgrades when using tendermint consensus
+  - There is a bug that is detected that could halt the chain but hasn't yet. There are other things we can develop to combat such scenarios. One thing we can do is develop a circuit breaker similar to the designs proposed in [Cosmos SDK](https://github.com/cosmos/cosmos-sdk/tree/main/x/circuit). This can disable certain message types or modules either in `CheckTx` or `ProcessProposal`. This violates the consistency property between `PrepareProposal` and `ProcessProposal` but so long as a quorum are the same, will still allow the chain to progress (inconsistency here can be interpreted as byzantine).
 
-### Phase 1: Configured Upgrade Height
+### Future Work: Signaled Upgrade Height
 
-The height of the upgrades will initially be coordinated via the `app.toml` config file under a seprate upgrades section. This will consist of a mapping from chain ID to height to app version that will be loaded by the application into working memory whenever the node begins. The `upgrades` module will simply take in this map and a reference to the `ParamStore` which it can use to set the new app version at the appropriate height. For safety, users will not be able to specify an app version that is greater than what the binary supports (i.e. 10 for v8). There is no rule preventing users from specifying a downgrade to an older version or a version that skips values. For convenience to node operators, a default mapping can be included in the binary such that the node operators simply need to stop the node, download the appropriate binary and restart the node.
+Preconfigured upgrade paths are vulnerable to halts. There is no indication that a quorum has in fact upgraded and that when the proposer proposes a block with the message to change version, that consensus will be reached. To mitigate this risk, the upgrade height can instead be signaled by validators. A version of `VoteExtension`s may be the most effective at ensuring this. Validators upon start up will automatically signal a version upgrade when they go to vote (i.e. `ExtendedVote`) so long as the latest supported version differs from the current network version. In `VerifyVoteExtension`, the version will be parsed and persisted (although not part of state). There is no verification. Upon a certain threshold which must be at least 2/3+ but could possibly be greater, the next proposer, who can support this version will propose a block with the `MsgVersionChange` that the quorum have agreed to. The rest works as before.
```

Comment:
imo, the only time anyone should use vote extensions should be to add constraints to block production. aka a block can only be produced if consensus is reached on ______.

1.) do you think that is the correct way to think about this here?
2.) would it be impossible to implement a system that has the same properties as what is described here either by using a separate gossiping mechansim or a stateful one like what is described below?

Reply:
Yes, it's possible as I mention further down to have a separate gossiping mechanism for signalling. The reason for favouring `VoteExtensions` over this is that there is already support for the signer to sign these arbitrary bytes plus all the logic to verify the signatures. 

I do think the separate process would be cleaner perhaps